### ISSUE TITLE
PowerShell attach removes existing breakpoints in attached session

### ISSUE THREAD

**Issue by nightroman** [PowerShell/vscode-powershell#4489]
### Prerequisites

- [X] I have written a descriptive issue title.
- [X] I have searched all [_open and closed_ issues](https://github.com/PowerShell/vscode-powershell/issues?q=is%3Aissue) to ensure it has not already been reported.
- [X] I have read the [troubleshooting](https://github.com/PowerShell/vscode-powershell/blob/main/docs/troubleshooting.md) guide.
- [X] I am sure this issue is with the _extension itself_ and does not reproduce in a standalone [PowerShell](https://github.com/PowerShell/PowerShell/issues/new/choose) instance.
- [X] I have verified that I am using the latest version of Visual Studio Code and the PowerShell extension.
- [X] If this is a security issue, I have read the [security issue reporting guidance](https://github.com/PowerShell/vscode-powershell/blob/main/SECURITY.md).

### Summary

I use VSCode PowerShell debugger by attaching to a process hosting PowerShell.
Attaching removes breakpoints that I have already set in my session.

### PowerShell Version

```console
Name                           Value
----                           -----
PSVersion                      7.3.3
PSEdition                      Core
GitCommitId                    7.3.3
OS                             Microsoft Windows 10.0.19045
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

### Visual Studio Code Version

```console
1.73.1
6261075646f055b99068d3688932416f2346dd3b
x64
```

### Extension Version

```console
ms-vscode.powershell@2023.3.3
```

### Steps to Reproduce

(1) Invoke the below script `test1.ps1` in some PowerShell session. E.g. I used and saw the problem in powershell and pwsh consoles and in my application hosting PowerShell as well.

```powershell
function global:test1 {
	'hi'
}
Set-PSBreakpoint -Command test1
Set-PSBreakpoint -Script $PSScriptRoot\test1.ps1 -Line 2
```

(2) Run `Get-PSBreakpoint` -- see 2 breakpoints

(3) In VSCode, launch/attach PowerShell debugger configured as below; select the target process

```json
      {
        "type": "PowerShell",
        "request": "attach",
        "name": "PowerShell attach",
        "processId": "${command:PickPSHostProcess}",
        "runspaceId": 1
      },
```

(4) With the debugger attached, run `Get-PSBreakpoint` again -- see that breakpoints are gone.

**Comment by nightroman** [comment]
@andschwa Noted, thank you. Looking forward to the progress and improvements.

FWIW, I believe (strongly) it used to work fine, breakpoints were not removed. That was my usual workflow - set breakpoints in the app, then start VSCode debugging. Granted, it was some time ago. Maybe newer versions are "not supposed to work" due to their new tech issues.

**Comment by andschwa** [comment]
Hard to say if/when the behavior changed. As far as I remember the debugger has always issued a "clear breakpoints" command on startup, but it's a project with a looong history.

**Comment by nightroman** [comment]
Some sort of prove, my user manual -- https://github.com/nightroman/FarNet/tree/main/PowerShellFar#debugging

> Select this debugger as active and start (F5). In the shown list of processes hosting PowerShell select Far Manager. Assuming you have set some breakpoints (**in Far Manager, not VSCode**), run your PowerShellFar code. If breakpoints are hit then VSCode debugger opens the source code at the active breakpoint.

**Comment by nightroman** [comment]
Updated the manual and gave the users some hope :)

[#4489]: https://github.com/PowerShell/vscode-powershell/issues/4489

> With the latest VSCode and PowerShell breakpoints set in Far Manager are
removed when VSCode debugger attaches. If you experience this issue, set
breakpoints in VSCode instead. Hopefully, this issue will be fixed, [#4489].

**Comment by SeeminglyScience** [comment]
Yes I think at one point it "worked" in that it just didn't try to manage breakpoints in the remote session at all. Or at most just repeatedly set the same breakpoints every time you attached. I know it's not the best for your scenario but the old behavior leads to some pretty wonky UX if you are trying to set breakpoints in the UI.

As Andy said though, #4065 should make the behavior more consistent (and I'll be returning to it shortly). I don't think I touch existing breakpoints in remote sessions.

**Comment by nightroman** [comment]
FWIW, PowerShellFar (Far Manager PS host) takes care of breakpoint updates via events, i.e. breakpoints set by a user in UI, or by commands, or as I can see even set in VSCode after attaching the debugger.

**Comment by nightroman** [comment]
Anyway, it's good to know that issue is known and addressed, one way or another. I'll adjust.

**Comment by nightroman** [comment]
Hello, with the recent development on fixing the debugger (https://github.com/PowerShell/vscode-powershell/releases/tag/v2024.3.0-preview), is there any chance this issue gets some progress?

**Comment by andyleejordan** [comment]
No, that work is happening with https://github.com/PowerShell/vscode-powershell/pull/4065 which is still in progress.

**Comment by nightroman** [comment]
FWIW, not complaining about anything, but please take a look at the manual for my users on using VSCode PowerShell debugger. You might and should be interested in the section "Known issues" (4). It's a lot of pain to make working apart from configuring, and IMHO many users will never enjoy VSCode PowerShell debugging by attaching (which actually works at this point, wow!). 

https://github.com/nightroman/FarNet/wiki/PowerShellFar-debugging

**Comment by Kaled1992** [comment]
{
        "type": "PowerShell",
        "request": "attach",
        "name": "PowerShell attach",
        "processId": "${command:PickPSHostProcess}",
        "runspaceId": 1
      },

**Comment by Kaled1992** [comment]
{
    "type": "PowerShell",
    "request": "attach",
    "name": "PowerShell attach",
    "processId": "${command:PickPSHostProcess}",
    "runspaceId": 1
  },

**Comment by nightroman** [comment]
Hello. Is there any chance the issue is fixed?
I'm about to close the issue as not been fixed.

**Comment by JustinGrote** [comment]
Not fixed yet but not forgotten! :)

**Comment by nightroman** [comment]
And I cannot tell my users "Do not set breakpoints in this IDE script editors, do this in VSCode on attaching its debugger".
Because VSCode debugger is not the only available and used, users choose which one to use depending on a case.
This debugger is used as well, it assumes pre-set breakpoints -- [Add-Debugger.ps1](https://github.com/nightroman/PowerShelf/blob/main/Add-Debugger.ps1)

**Comment by JustinGrote** [comment]
@nightroman so what I think we might be able to do at least in this case is interrogate the session during an attach for existing breakpoints and use the [Breakpoint Event](https://microsoft.github.io/debug-adapter-protocol/specification#Events_Breakpoint) to inform the "client" of those existing breakpoints, so we don't have to reset them. The DAP spec is not clear when these should be sent but it looks like sending it anytime after `configurationDone` is acceptable, so we could send after the `attach` request and either just before or immediately after the acknowledgement. I'm unsure how various DAP clients (vscode, vim, etc.) would behave so we'd have to test it.

I don't have any current plans on my roadmap to implement this to be honest, as a volunteer I just have some other project priorities at the moment.
